bin/client/src/precompiles/bls12.rs

```diff
@@ -0,0 +1,115 @@
+//! Contains the accelerated precompile for the BLS12-381 curve.
+//!
+//! BLS12-381 is introduced in [EIP-2537](https://eips.ethereum.org/EIPS/eip-2537).
+//!
+//! For constants and logic, see the [revm implementation].
+//!
+//! [revm implementation]: https://github.com/bluealloy/revm/blob/main/crates/precompile/src/bls12_381/pairing.rs
+
+use crate::{HINT_WRITER, ORACLE_READER};
+use alloc::{string::ToString, vec::Vec};
+use alloy_primitives::{address, keccak256, Address, Bytes};
+use kona_preimage::{
+    errors::PreimageOracleError, HintWriterClient, PreimageKey, PreimageKeyType,
+    PreimageOracleClient,
+};
+use kona_proof::{errors::OracleProviderError, HintType};
+use revm::{
+    precompile::{Error as PrecompileError, Precompile, PrecompileResult, PrecompileWithAddress},
+    primitives::PrecompileOutput,
+};
+
+/// The address of the BLS12-381 pairing check precompile.
+const BLS12_PAIRING_CHECK: Address = address!("0x000000000000000000000000000000000000000f");
+
+/// Input length of pairing operation.
+const INPUT_LENGTH: usize = 384;
+
+/// Multiplier gas fee for BLS12-381 pairing operation.
+const PAIRING_MULTIPLIER_BASE: u64 = 32600;
+
+/// Offset gas fee for BLS12-381 pairing operation.
+const PAIRING_OFFSET_BASE: u64 = 37700;
+
+/// The address of the BLS12-381 pairing precompile.
+pub(crate) const FPVM_BLS12_PAIRING: PrecompileWithAddress =
+    PrecompileWithAddress(BLS12_PAIRING_CHECK, Precompile::Standard(fpvm_bls12_pairing));
+
+/// Performs an FPVM-accelerated BLS12-381 pairing check.
+fn fpvm_bls12_pairing(input: &Bytes, gas_limit: u64) -> PrecompileResult {
+    let input_len = input.len();
```

Comment:
We need to make a diff to the L1 precompile that limits the amount of gas it can consume, similar to what we did in granite for bn128 pairing. Have we spec'd this yet?

Reply:
e: i see the discussion over in https://github.com/ethereum-optimism/specs/pull/553/files#r1931103051, looks like the right direction. This is good to go once that limit has been put in place in both the accelerated precompile and op-revm.

Reply:
is there a way to catch this in a test?

Reply:
Oh good callout - this is not spec'd out yet thanks